File: src/Other/Internals/Menu/UIElementHelpers.vb
Imports System.Windows

Namespace OpenSilver.Samples.Showcase
    Public NotInheritable Class UIElementHelpers
        Public Shared Function WaitForLoadedAsync(element As FrameworkElement) As Task
            If element.IsLoaded Then
                Return Task.CompletedTask
            End If

            Dim tcs As New TaskCompletionSource(Of Object)()
            Dim handler As RoutedEventHandler = Nothing

            handler = Sub(s, e)
                          RemoveHandler element.Loaded, handler
                          tcs.SetResult(Nothing)
                      End Sub

            AddHandler element.Loaded, handler
            Return tcs.Task
        End Function
    End Class
End Namespace

File: src/OpenSilver.Showcase/Other/Internals/Menu/UIElementHelpers.fs
namespace OpenSilver.Showcase

open System.Threading.Tasks
open System.Windows

type UIElementHelpers() =
    static member WaitForLoadedAsync(element: FrameworkElement) : Task =
        if element.IsLoaded then
            Task.CompletedTask
        else
            let tcs = TaskCompletionSource<unit>()
            let mutable handler : RoutedEventHandler = Unchecked.defaultof<_>
            handler <- RoutedEventHandler(fun _ _ ->
                element.Loaded.RemoveHandler(handler)
                tcs.SetResult(())
            )
            element.Loaded.AddHandler(handler)
            tcs.Task :> Task
